wf/clean.py

```diff
@@ -0,0 +1,172 @@
+    
+    apply_medians_c = singlecell.loc[:,'col'].apply(lambda x: all_indexes_c[x])
+
+    row_mean = statistics.mean(list(all_indexes_r.values()))
+    row_std = statistics.stdev(list(all_indexes_r.values()))
+    col_mean = statistics.mean(list(all_indexes_c.values()))
+    col_std = statistics.stdev(list(all_indexes_c.values()))
+    # identify limit more than x standard deviations above mean
+    rows_limit = row_mean + deviations * row_std
+    cols_limit = col_mean + deviations * col_std
+
+    # create table with only diagonal tixels from singlecell table
+    diag_sc = singlecell[singlecell['row'] == singlecell['col']]
+    diag_mean = diag_sc['passed_filters'].mean()
+
+    # create 'adjust' column with reads to downsample
+    if diag_mean > rows_limit:
+        diag_sc['adjust'] = np.ceil(diag_sc['passed_filters'] * (row_mean / diag_mean))
+    elif diag_mean > cols_limit:
+        diag_sc['adjust'] = np.ceil(diag_sc['passed_filters'] * (col_mean / diag_mean))
+    else:
+        diag_sc['adjust'] = diag_sc['passed_filters']
+
+    return diag_sc[['barcode', 'adjust']]
+
+def combine_tables(singlecell, deviations=1):
+    row_singlecell = singlecell.copy()
+    col_singlecell = singlecell.copy()
+    dia_singlecell = singlecell.copy()
+    row_reductions = get_reductions(row_singlecell, 'row', deviations)
+    col_reductions = get_reductions(col_singlecell, 'col', deviations)
+    diag_reductions = get_diag_reductions(dia_singlecell, deviations)
+
+    # concat rows and columns
+    # If a tixel occurs twice, take the average value
+    combined_table = averageDuplicates([row_reductions.values.tolist(), col_reductions.values.tolist(), diag_reductions.values.tolist()])
+    
+    return combined_table
+
+def clean_fragments(fragments_path, r_table):
+    # Reduce high tixels by randomly downsampling fragments.tsv according to
+    # reduction table
+    skip_headers = [i+1 for i in range(50)]
+    print("Loading fragments.tsv")
+    # fragments = gzip.open(fragments_path)
+    fragments = pd.read_csv(fragments_path, sep='\t', header=skip_headers, skiprows=skip_headers)
+    fragments.columns = ['V1', 'V2', 'V3', 'barcode', 'V4']
+    frag_copy = fragments.copy()
+    outlier_barcodes = list(r_table.keys())
+    fragments = fragments[fragments['barcode'].isin(outlier_barcodes) == False]
+    frag_copy = frag_copy[frag_copy['barcode'].isin(outlier_barcodes) == True]
+    # To each df in the list, randomly downsample if in reduction list
+    print("Downsampling....")
+    barcode_groups = frag_copy.groupby('barcode')
+    list_concat = []
+    for i in outlier_barcodes:
+      unique_barcode = barcode_groups.get_group(i)
+      if unique_barcode.shape[0] > int(r_table[i]):
+        unique_barcode = unique_barcode.sample(n= int(r_table[i]))
+      list_concat.append(unique_barcode)
+    
+    downsampled_barcode = pd.concat(list_concat)
+    fragments_cleaned = pd.concat([downsampled_barcode, fragments])
```

Comment:
This is a clever way to do this.  We don't need the copy though.  Let's try,
```suggestion
    outlier_barcodes = list(r_table.keys())
    normal_frags = fragments[fragments['barcode'].isin(outlier_barcodes) == False]
    outlier_frags = frag_copy[frag_copy['barcode'].isin(outlier_barcodes) == True]
    
    # To each df in the list, randomly downsample if in reduction list
    print("Downsampling....")
    barcode_groups = outlier_frags.groupby('barcode')
    list_concat = []
    for i in outlier_barcodes:
      outlier = barcode_groups.get_group(i)
      if outlier.shape[0] > int(r_table[i]):
        outlier = outlier.sample(n= int(r_table[i]))
      list_concat.append(outlier)
    
    downsampled_frags = pd.concat(list_concat)
    fragments_cleaned = pd.concat([downsampled_frags, normal_frags])
```